### ISSUE TITLE
Streaming request returns 401 authentication required

### ISSUE THREAD

**Issue by nickdtodd** [karlheyes/icecast-kh#130]
Hi Karl,

This is possibly a duplicate of issue https://github.com/karlheyes/icecast-kh/issues/100

Using the current build: `21-g270abdb` with logging set to debug level.

This would appear to be an intermittent fault. After some hours with no issues we started getting 401's from many, if not all, mountpoints.

Occasionally there was a 200 among the 401's but they were few and far between. You can see this from the [access log](http://nopaste.linux-dev.org/?931904).

Here are two error logs that are from around the time of the problem:

[before/during error log](http://nopaste.linux-dev.org/?931903)
[after error log](http://nopaste.linux-dev.org/?931905)

These are concurrent logs and there is a gap where there was no logging to the error log. The period where the 401's are occurring do not have error.log entries at that time.

The times between the logging stopping and starting:
```
[2016-01-18  16:40:14] <-- last entry in log error.log.20160118_164014
[2016-01-18  17:10:12] <-- first entry in log error.log.20160118_180104
```

It would appear that Icecast never went down. I've checked all the usual system logs, syslog, dmesg, messages, etc. but there is no segfault or anything nasty like that.

Here is a New Relic chart showing the Icecast process at this time.

![screen shot 2016-01-19 at 10 22 11](https://cloud.githubusercontent.com/assets/3327975/12417619/3c84ceba-bea1-11e5-87f5-f92f751b3865.png)

If you need anything else please ask.

Thanks, Nick

**Comment by karlheyes** [comment]
seems suspiciously precise to 30 mins of logging outage. The naming would indicate log cycling going on, you obviously have url auth in play, 2 workers and very close to 400 clients.  Have you set any limits on clients/max-listeners/max-bandwidth or ulimit?

A logging stop is odd indeed, check prlimit -p pidof icecast or cat /proc/<pid>/limits and how many entries there are in lsof -pPp <pid>

karl.

**Comment by nickdtodd** [comment]
Hi Karl,

There is a limit of `10000` set in the icecast config for clients, nothing for bandwidth limits.

Here is some other details that you asked for:

```
prlimit -p $(pidof icecast)

RESOURCE   DESCRIPTION                             SOFT      HARD UNITS
AS         address space limit                unlimited unlimited bytes
CORE       max core file size                 unlimited unlimited blocks
CPU        CPU time                           unlimited unlimited seconds
DATA       max data size                      unlimited unlimited bytes
FSIZE      max file size                      unlimited unlimited blocks
LOCKS      max number of file locks held      unlimited unlimited
MEMLOCK    max locked-in-memory address space     65536     65536 bytes
MSGQUEUE   max bytes in POSIX mqueues            819200    819200 bytes
NICE       max nice prio allowed to raise             0         0
NOFILE     max number of open files               10000     10000
NPROC      max number of processes                64159     64159
RSS        max resident set size              unlimited unlimited pages
RTPRIO     max real-time priority                     0         0
RTTIME     timeout for real-time tasks        unlimited unlimited microsecs
SIGPENDING max number of pending signals          64159     64159
STACK      max stack size                       8388608 unlimited bytes
```
That value for `NOFILE` seems at odds to what has been set for the user running icecast:

```
ulimit -a

core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 64159
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 1000000
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 64159
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
```

Next:
```
lsof -p $(pidof icecast) | wc -l
70
```

I think this low due to the fact that there is no traffic being sent to this server while we figure out what's going on here. On another server, on an older build version, that has active connections came out with a count of `~810`

Anything else, please ask.

Thanks, Nick.

**Comment by karlheyes** [comment]
The prlimit shows 10000 as a limit so you should be ok there, especially if you are within 1000 connections.

I'm still not sure on the events yet, the first thing is to get the error log working as it should, I had thought a FD limit maybe breaking the log cycle but that does not seem to be the case.  The 401 could be down to a stalled auth, eg libcurl still requires file handles, but the logging question still needs to be resolved.

What really need to be done is during this logging outage, find out what it is doing, an lsof -nPp pid  to see what file is actually open for logging (is it deleted, renamed etc), and possible an strace of the process (strace -o output -tt -f -p pid).  There will be regular stats reporting for any active stream so something should be tried even if it fails.

karl.

**Comment by nickdtodd** [comment]
Thanks Karl.  We'll give it some traffic and see if we can get some more information.

I'll report here when I know more.

Thanks, Nick.

**Comment by nickdtodd** [comment]
Hi Karl,

We ran into 401's again now. I ran `strace` while we were experiencing the issues:

```
root@host492096:~# strace -o output -tt -f -p $(pidof icecast)
Process 3638 attached with 9 threads
Process 5260 attached
Process 5261 attached
Process 5262 attached
Process 5263 attached
Process 5264 attached
Process 5268 attached
Process 5269 attached
Process 5270 attached
Process 5271 attached
Process 5272 attached
Process 5273 attached
Process 5274 attached
Process 5275 attached
Process 5276 attached
Process 5277 attached
Process 5278 attached
Process 5279 attached
Process 5280 attached
Process 5281 attached
Process 5282 attached
Process 5283 attached
Process 5284 attached
Process 5285 attached
Process 5286 attached
Process 5287 attached
Process 5288 attached
Process 5289 attached
Process 5290 attached
```
This is the output from running `lsof -nPp $(pidof icecast)`: http://nopaste.linux-dev.org/?931934

Here is the `error log` for the period: http://nopaste.linux-dev.org/?931932
And the `access log`: http://nopaste.linux-dev.org/?931935

One other point of interest - while it was happening, and you tried to hit a stream in the browser, it would render a load of text on some occasions, the standard http access login box on others.

Thanks, Nick.

**Comment by karlheyes** [comment]
the lsof indicates the error log is not open, do you have any particular requirements on the size of the error log, I'm thinking the log cycle trigger is kicking in but why it's not managing the resume I'm not sure.  Can you show me the log settings you have.  we may need to simulate a log file cycle triggering.
Still not entirely sure how 401 may trigger. Is this in a chroot jail?

karl.

**Comment by nickdtodd** [comment]
Hi Karl,

Here is the logging and chroot config:

```
    <logdir>/home/server/logs</logdir>

    ...

    <logging>
        <accesslog>access.log</accesslog>
        <errorlog>error.log</errorlog>
        <!-- <playlistlog>playlist.log</playlistlog> -->
        <loglevel>4</loglevel> <!-- 4 Debug, 3 Info, 2 Warn, 1 Error -->
        <logsize>10000</logsize> <!-- Max size of a logfile -->
        <!-- If logarchive is enabled (1), then when logsize is reached
             the logfile will be moved to [error|access|playlist].log.DATESTAMP,
             otherwise it will be moved to [error|access|playlist].log.old.
             Default is non-archive mode (i.e. overwrite)
        -->
        <!-- <logarchive>1</logarchive> -->
    </logging>

    <security>
        <chroot>0</chroot>
        <changeowner>
            <user>deploy</user>
            <group>deploy</group>
        </changeowner>
    </security>
```
If there is anything else you want me to do please let me know.

Thanks, Nick.

**Comment by karlheyes** [comment]
On another report, there was a case identified where a file handle was closed when it should not of been. In that case it affected a worker communication channel but it could in another situation affect other open files (like log files) and the trigger would be something like a 404 response.   I've now fixed that in the master so that should solve the other bug and It could very well fix this or at least allow the logging to continue

karl.
